### DISCUSSION TITLE
Welcome to BindDiffusion Discussions!

### DISCUSSION THREAD

**Discussion by yzhwang** [sail-sg/BindDiffusion#3] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by matthewachan** [comment]
Hi, thanks for the great work on this!

I went through this [Jupyter notebook](https://github.com/sail-sg/BindDiffusion/blob/46ea7e3f2d9b7a8e739a876e2deb3cebad98c1a7/audio2image.ipynb) in a previous commit, and I'm still a little confused about how your approach works. Any explanation would be appreciated!

The way I currently understand it, ImageBind learns a joint latent space for different modalities (i.e. audio, image, text, etc.) and [the pretrained model](https://github.com/facebookresearch/ImageBind) provided by Meta allows us to encode data into this latent space.

What I'm confused about is how you're conditioning a pre-trained LDM on ImageBind encoded image, audio, text, etc. latents and then decoding those results. In particular, I'm not really clear on what's happening in this code excerpt from the Jupyter notebook I mentioned (especially the lines after the call to `noise_augmentor`):
```
with torch.no_grad(), torch.autocast('cuda'):
    c_adm, noise_level_emb = model.noise_augmentor(c_adm, noise_level=torch.zeros(batch_size).long().to(c_adm.device))
    # assume this gives embeddings of noise levels
    c_adm = torch.cat((c_adm, noise_level_emb), 1)

    uc = model.get_learned_conditioning(batch_size * [""])
    uc = {"c_crossattn": [uc], "c_adm": torch.zeros_like(c_adm)}
    c = {"c_crossattn": [model.get_learned_conditioning(prompts)], "c_adm": c_adm}

    if mode == "editing":
        init_image = load_img(init_img).to(device)
        init_latent = model.get_first_stage_encoding(model.encode_first_stage(init_image))  # move to latent space
        z_enc = sampler.stochastic_encode(init_latent, torch.tensor([t_enc] * batch_size).to(device))

        # decode it
        samples = sampler.decode(z_enc, c, t_enc, unconditional_guidance_scale=opt.scale,
                                unconditional_conditioning=uc, )
    else:
        samples, _ = sampler.sample(S=opt.steps,
                                    conditioning=c,
                                    batch_size=opt.n_samples,
                                    shape=shape,
                                    verbose=False,
                                    unconditional_guidance_scale=opt.scale,
                                    unconditional_conditioning=uc,
                                    eta=opt.ddim_eta,
                                    x_T=start_code)

x_samples = model.decode_first_stage(samples)
x_samples = torch.clamp((x_samples + 1.0) / 2.0, min=0.0, max=1.0)
```

**Reply by ikuinen** [reply]
Thanks for your interest and great question!

The pretrained checkpoint can take a image embedding as extra condition(i.e. _c_adm_) and produce images with similar semantics(i.e. image variations). The embedding is added to the time embedding after mapping. During training, the image embedding is also corrupted in a given noise level(i.e. noise augment). Therefore, during inference, the model allows a bigger variation around the semantics of images if a higher noise level is given.

**Reply by unography** [reply]
The pre-trained model is stability's unCLIP model which has a separate training than ImageBind

Are you saying that you're able to match unclip's embedding ~ imagebind's embedding + noise?

**Reply by ikuinen** [reply]
> The pre-trained model is stability's unCLIP model which has a separate training than ImageBind
> 
> Are you saying that you're able to match unclip's embedding ~ imagebind's embedding + noise?

yeah, as you said. Both the embedding is actually aligned with clip image embedding.

**Comment by datovar4** [comment]
Amazing work so far! Any plans to add functionality to go from image to audio or text to audio?

**Reply by ikuinen** [reply]
> Amazing work so far! Any plans to add functionality to go from image to audio or text to audio?

Yeah, working on it
